recursion/core/src/poseidon2_wide/external.rs

```diff
@@ -89,204 +413,129 @@ where
-        //     //
-        //     // We first precompute the four sums of every four elements.
-        //     let sums: [AB::Expr; 4] = core::array::from_fn(|k| {
-        //         (0..WIDTH)
-        //             .step_by(4)
-        //             .map(|j| state[j + k].clone())
-        //             .sum::<AB::Expr>()
-        //     });
-
-        //     // The formula for each y_i involves 2x_i' term and x_j' terms for each j that equals i mod 4.
-        //     // In other words, we can add a single copy of x_i' to the appropriate one of our precomputed sums.
-        //     for i in 0..WIDTH {
-        //         state[i] += sums[i % 4].clone();
-        //         builder
-        //             .when(local.is_external + local.is_initial)
-        //             .assert_eq(state[i].clone(), local.output[i]);
-        //     }
-        // }
-
-        // // INTERNAL LAYER
-        // {
-        //     // Use a simple matrix multiplication as the permutation.
-        //     let mut state: [AB::Expr; WIDTH] = sbox_result.clone();
-        //     let matmul_constants: [<<AB as AirBuilder>::Expr as AbstractField>::F; WIDTH] =
-        //         MATRIX_DIAG_16_BABYBEAR_U32
-        //             .iter()
-        //             .map(|x| <<AB as AirBuilder>::Expr as AbstractField>::F::from_wrapped_u32(*x))
-        //             .collect::<Vec<_>>()
-        //             .try_into()
-        //             .unwrap();
-        //     matmul_internal(&mut state, matmul_constants);
-        //     for i in 0..WIDTH {
-        //         builder
-        //             .when(local.is_internal)
-        //             .assert_eq(state[i].clone(), local.output[i]);
-        //     }
-        // }
-
-        // // Range check all flags.
-        // for i in 0..local.rounds.len() {
-        //     builder.assert_bool(local.rounds[i]);
-        // }
-        // builder.assert_bool(local.is_initial);
-        // builder.assert_bool(local.is_external);
-        // builder.assert_bool(local.is_internal);
-        // builder.assert_bool(local.is_initial + local.is_external + local.is_internal);
-
-        // // Constrain the initial flag.
-        // builder.assert_eq(local.is_initial, local.rounds[0]);
-
-        // // Constrain the external flag.
-        // let is_external_first_half = (0..4).map(|i| local.rounds[i + 1].into()).sum::<AB::Expr>();
-        // let is_external_second_half = (26..30)
-        //     .map(|i| local.rounds[i + 1].into())
-        //     .sum::<AB::Expr>();
-        // builder.assert_eq(
-        //     local.is_external,
-        //     is_external_first_half + is_external_second_half,
-        // );
+        // internal rounds
+        // first we do 7 chunks of 3 internal rounds
+        // then we do the last internal round
+        for chunk in 0..NUM_INTERNAL_ROUND_CHUNKS {
+            Self::build_internal_round_chunk(builder, cols, chunk);
```

Comment:
instead of build, can you use "eval_" 

also, coding style-wise, we typically just implement helper eval functions as raw functions instead of impl's cuz they don't take self.